app/controllers/users_controller.rb

```diff
@@ -6,13 +6,34 @@ def new
     @user = User.new
   end
 
+  def new_with_invite
+    @invitation = find_invite(params[:token])
+
+    if @invitation
+      @user = User.new(email_address: @invitation.friend_email)
+      render :new
+    else
+      redirect_to invalid_token_path
+    end
+  end
+
   def create
     @user = User.new(user_params)
 
     if @user.save
+      invite_token = params[:user][:invite_token]
       flash[:success] = "You have been registered successfully!"
       session[:user_id] = @user.id
       UserMailer.welcome_email(@user).deliver
+
+      if invite_token.present?
+        invite = find_invite(invite_token)
+        @inviter = invite.inviter
+        current_user.follow(@inviter)
```

Comment:
I think `current_user` could easily be `@user` and the reason for this is within the same action you are pouting to the same user but `current_user` must reach out to database once more. Remember prior here you haven't used it. Also mixing who is who it is not clear - I have to check up you signed him in `session[:user_id] = @user.id`.